File: backend/app/ports.py
from __future__ import annotations

from collections.abc import Iterator
from typing import Protocol

from app.domain.models import (
    ChatMessage,
    ExtractionResult,
    GeneratedCopy,
    MarketingEmail,
    ProductBrief,
    Violation,
)


class LLMClient(Protocol):
    def extract(
        self,
        message: str,
        brief: ProductBrief,
        history_tail: list[ChatMessage],
    ) -> ExtractionResult: ...

    def stream_product_description(self, brief: ProductBrief) -> Iterator[str]: ...

    def stream_email_body(
        self, brief: ProductBrief, product_description: str
    ) -> Iterator[str]: ...

    def generate_email(
        self,
        brief: ProductBrief,
        product_description: str,
        *,
        body: str,
    ) -> MarketingEmail: ...

    def repair(
        self,
        brief: ProductBrief,
        previous_output: GeneratedCopy,
        violations: list[Violation],
    ) -> GeneratedCopy: ...

File: backend/src/ports.ts
import type {
  ChatMessage,
  ExtractionResult,
  GeneratedCopy,
  MarketingEmail,
  ProductBrief,
  Violation,
} from "./domain/models.ts";

export interface LLMClient {
  extract(
    message: string,
    brief: ProductBrief,
    historyTail: ChatMessage[],
  ): Promise<ExtractionResult>;

  streamProductDescription(brief: ProductBrief): AsyncIterable<string>;

  streamEmailBody(
    brief: ProductBrief,
    productDescription: string,
  ): AsyncIterable<string>;

  generateEmail(
    brief: ProductBrief,
    productDescription: string,
    options: { body: string },
  ): Promise<MarketingEmail>;

  repair(
    brief: ProductBrief,
    previousOutput: GeneratedCopy,
    violations: Violation[],
  ): Promise<GeneratedCopy>;
}
